67_RayQueryGeometry/main.cpp

```diff
@@ -887,36 +910,37 @@ class RayQueryGeometryApp final : public examples::SimpleWindowedApplication, pu
 			}
 
 			// compact blas
-			IQueryPool::SCreationParams qParams{ .queryCount = 1, .queryType = IQueryPool::ACCELERATION_STRUCTURE_COMPACTED_SIZE };
-			smart_refctd_ptr<IQueryPool> queryPool = m_device->createQueryPool(std::move(qParams));
-
-			uint32_t queryCount = 0;
-			const IGPUAccelerationStructure* ases[1u] = { gpuBlas.get() };
-			cmdbuf->writeAccelerationStructureProperties({ ases, 1}, IQueryPool::ACCELERATION_STRUCTURE_COMPACTED_SIZE, queryPool.get(), queryCount++);
-
-			size_t asSizes[1];
-			m_device->getQueryPoolResults(queryPool.get(), 0, queryCount, asSizes, sizeof(size_t), IQueryPool::WAIT_BIT);
-			
-			auto cleanupBlas = gpuBlas;
-			{
-				IGPUBuffer::SCreationParams params;
-				params.usage = bitflag(IGPUBuffer::EUF_SHADER_DEVICE_ADDRESS_BIT) | IGPUBuffer::EUF_ACCELERATION_STRUCTURE_STORAGE_BIT;
-				params.size = asSizes[0];
-				smart_refctd_ptr<IGPUBuffer> asBuffer = createBuffer(params);
-
-				IGPUBottomLevelAccelerationStructure::SCreationParams blasParams;
-				blasParams.bufferRange.buffer = asBuffer;
-				blasParams.bufferRange.offset = 0u;
-				blasParams.bufferRange.size = asSizes[0];
-				blasParams.flags = IGPUBottomLevelAccelerationStructure::SCreationParams::FLAGS::NONE;
-				gpuBlas = m_device->createBottomLevelAccelerationStructure(std::move(blasParams));
-			}
-
-			IGPUBottomLevelAccelerationStructure::CopyInfo copyInfo;
-			copyInfo.src = cleanupBlas.get();
-			copyInfo.dst = gpuBlas.get();
-			copyInfo.mode = IGPUBottomLevelAccelerationStructure::COPY_MODE::COMPACT;
-			cmdbuf->copyAccelerationStructure(copyInfo);
+			//IQueryPool::SCreationParams qParams{ .queryCount = 1, .queryType = IQueryPool::ACCELERATION_STRUCTURE_COMPACTED_SIZE };
+			//smart_refctd_ptr<IQueryPool> queryPool = m_device->createQueryPool(std::move(qParams));
+			//cmdbuf->resetQueryPool(queryPool.get(), 0, 1);
+
+			//uint32_t queryCount = 0;
+			//const IGPUAccelerationStructure* ases[1u] = { gpuBlas.get() };
+			//cmdbuf->writeAccelerationStructureProperties({ ases, 1}, IQueryPool::ACCELERATION_STRUCTURE_COMPACTED_SIZE, queryPool.get(), queryCount++);
+
+			//size_t asSizes[1];
+			//m_device->getQueryPoolResults(queryPool.get(), 0, queryCount, asSizes, sizeof(size_t), IQueryPool::WAIT_BIT);
```

Comment:
you were using the query wrong, if something is a command buffer operation (`IGPUCommandBuffer` method) that means that it completes when the command buffer it was recorded into finishes execution.

You can only get the Query Pool Results AFTER the semaphore completes.

This would mean that you would need 2 separate command buffers and queue submits with the Host blocking after each:
1. build blas
2. compact blas & build tlas

the good news is that you don't need a pipeline barrier between 1 and 2 because they're separated by a timeline semaphore signal-wait pair (indirectly, via the host)

Reply:
what you're doing now is actually causing a deadlock cause you're making the Host wait on the result of a query thats not even pending yet

Reply:
also its not necessary to reset a query pool before using it (more like After0